File: exercise4/Bionomial_probability/Bionomial_probability/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
namespace Bionomial_probability
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }
        private static long Factorial(long x)
        {
            if (x <= 1)
                return 1;
            else
                return x * Factorial(x - 1);
        }
        private static long Combination(long a, long b)
        {
            if (a <= 1)
                return 1;

            return Factorial(a) / (Factorial(b) * Factorial(a - b));
        }
        //Relative frequency of success
        //n-p-x
        public double BinomialProbability(int trials, int successes, double probabilityOfSuccess)
        {
            double probOfFailures = 1 - probabilityOfSuccess;

            
            Dictionary<int, double> probability_binomial = new Dictionary<int, double>();
            for (int i = 0; i <= successes; i++)
            {
                double c = Combination(trials, i);
                double px = Math.Pow(probabilityOfSuccess, i);
                double qnx = Math.Pow(probOfFailures, trials - i);
                probability_binomial.Add(i, (c * px * qnx));

                //plot
                

            }
            Plot_binomial();
            return probability_binomial.Values.Last();
        }
        public void Plot_binomial()
        {
            Chart chart = this.FindName("Chart") as Chart;
            if (chart != null)
            {
                chart.DataSource = probability_binomial;
                chart.Series["series"].XValueMember = "Key";
                chart.Series["series"].YValueMembers = "Value";
            }
        }
        private void textBox1_TextChanged(object sender, EventArgs e)
        {

        }

        private void btnCalculate_Click(object sender, EventArgs e)
        {
            int nbrOfTrials = 0, nbrOfSuccesses = 0, nbrOfDomain=0;
            double probOfSuccesses = 0.00;
            double binomial = 0.00;
            try
            {
                nbrOfDomain = int.Parse(textBox1_src.Text);
            }
            catch (FormatException)
            {
                MessageBox.Show("The number of domain is not valid", "Binomial Probability");
            }
            try
            {
                nbrOfTrials = int.Parse(txtTrials.Text);
            }
            catch (FormatException)
            {
                MessageBox.Show("The number of trials is not valid","Binomial Probability");
            }

            try
            {
                nbrOfSuccesses = int.Parse(txtSuccesses.Text);
            }
            catch (FormatException)
            {
                MessageBox.Show("The number of successes is not valid","Binomial Probability");
            }

            try
            {
                
               
                    probOfSuccesses = double.Parse(txtProbabilitySuccess.Text);

                    binomial = BinomialProbability(nbrOfTrials,
                                        nbrOfSuccesses,
                                        probOfSuccesses);
                    txtBinomialProbability.Text = binomial.ToString();

               
            }
            catch (FormatException)
            {
                MessageBox.Show("The number of trials is not valid","Binomial Probability");
            }
            
        }

        private void btnClose_Click(object sender, EventArgs e)
        {
            Close();
        }

        private void txtSuccesses_TextChanged(object sender, EventArgs e)
        {

        }

    

        private void textBox2_To(object sender, EventArgs e)
        {

        }

        private void label3_Click(object sender, EventArgs e)
        {

        }

        private void Form1_Load(object sender, EventArgs e)
        {

        }
    }
}

File: exercise4/binomial_distribution/binomial_distribution/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Public Class Form1
    Private Shared Function Factorial(ByVal x As Long) As Long
        If x <= 1 Then
            Return 1
        Else
            Return x * Factorial(x - 1)
        End If
    End Function
    Private Shared Function Combination(ByVal a As Long, ByVal b As Long) As Long
        If a <= 1 Then Return 1
        Return Factorial(a) / (Factorial(b) * Factorial(a - b))
    End Function
    Private Function BinomialProbability(ByVal trials As Integer, ByVal successes As Integer, ByVal probabilityOfSuccess As Double) As Double
        Dim probOfFailures As Double = 1 - probabilityOfSuccess
        Dim c As Double = Combination(trials, successes)
        Dim px As Double = Math.Pow(probabilityOfSuccess, successes)
        Dim qnx As Double = Math.Pow(probOfFailures, trials - successes)
        Return c * px * qnx
    End Function

    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles Label1.Click

    End Sub

    Private Sub Label2_Click(sender As Object, e As EventArgs) Handles Label2.Click

    End Sub

    Private Sub btnCalculate_Click(sender As Object, e As EventArgs) Handles btnCalculate.Click
        Dim nbrOfTrials As Integer = 0, nbrOfSuccesses As Integer = 0
        Dim probOfSuccesses As Double = 0.00
        Dim binomial As Double = 0.00

        Try
            nbrOfTrials = Integer.Parse(txtTrials.Text)
        Catch __unusedFormatException1__ As FormatException
            MessageBox.Show("The number of trials is not valid", "Binomial Probability")
        End Try

        Try
            nbrOfSuccesses = Integer.Parse(txtSuccesses.Text)
        Catch __unusedFormatException1__ As FormatException
            MessageBox.Show("The number of successes is not valid", "Binomial Probability")
        End Try

        Try
            probOfSuccesses = Double.Parse(txtProbabilitySuccess.Text)
            binomial = BinomialProbability(nbrOfTrials, nbrOfSuccesses, probOfSuccesses)
            txtBinomialProbability.Text = binomial.ToString()
        Catch __unusedFormatException1__ As FormatException
            MessageBox.Show("The number of trials is not valid", "Binomial Probability")
        End Try
    End Sub

    Private Sub btnClose_Click(sender As Object, e As EventArgs) Handles btnClose.Click
        Close()
    End Sub
End Class
